### ISSUE TITLE
TF_ROCM_FUSION_ENABLE fails on official/resnet/cifar10_main.py 

### ISSUE THREAD

**Issue by TheCBaH** [ROCmSoftwarePlatform/tensorflow-upstream#456]
**System information**
- Have I written custom code: No
- OS Platform and Distribution: Ubuntu 18.04 
- TensorFlow installed from: binary 
- TensorFlow version: 1,13
- Python version: 3.5
- ROCm/MIOpen version: 2.4
- GPU model and memory: gfx900/8GB

**Describe the current behavior**
When enabling TF_ROCM_FUSION_ENABLE  official/resnet/cifar10_main.py  model fails to run

**Describe the expected behavior**
TF_ROCM_FUSION_ENABLE should improve performan and do not fail

**Code to reproduce the issue**
```
$ docker run -it --rm --device=/dev/kfd --device=/dev/dri --group-add video  rocm/tensorflow:rocm2.4-tf1.13-python3
2326577ab9c8:~:$ cd /tmp/
@2326577ab9c8:/tmp$ git clone -b r1.13.0 --depth 1 https://github.com/tensorflow/models.git
Cloning into 'models'...
...
@2326577ab9c8:/tmp$ cd models/
@2326577ab9c8:/tmp/models$ TF_ROCM_FUSION_ENABLE=1 PYTHONPATH=`pwd`  python3.5 official/resnet/cifar10_download_and_extract.py
...
@2326577ab9c8:/tmp/models$ TF_ROCM_FUSION_ENABLE=1 PYTHONPATH=`pwd`  python3.5 official/resnet/cifar10_main.py

2019-05-15 02:28:48.760231: I tensorflow/core/graph/gpu_fusion_pass.cc:434] ROCm Fusion is enabled.
2019-05-15 02:28:48.767839: F tensorflow/core/graph/gpu_fusion_pass.cc:1346] Unexpected output connection on the BatchNormGrad node :id : 5696, name : gradients/resnet_model/batch_normalization/FusedBatchNorm_grad/FusedBatchNormGrad, type_string : FusedBatchNormGrad, num_inputs : 5, num_input_edges : 5, num_outputs : 5, num_output_edges : 6. output(idx, dest_node) :3, id : 5701, name : gradients/resnet_model/batch_normalization/FusedBatchNorm_grad/tuple/control_dependency_3, type_string : Identity, num_inputs : 1, num_input_edges : 2, num_outputs : 1, num_output_edges : 1
Fatal Python error: Aborted

Thread 0x00007ffe6affd700 (most recent call first):
  File "/usr/lib/python3.5/threading.py", line 293 in wait
  File "/usr/lib/python3.5/queue.py", line 164 in get
  File "/usr/local/lib/python3.5/dist-packages/tensorflow/python/summary/writer/event_file_writer.py", line 159 in run
  File "/usr/lib/python3.5/threading.py", line 914 in _bootstrap_inner
  File "/usr/lib/python3.5/threading.py", line 882 in _bootstrap
```
**Other info / logs**
resnet from benchmarks actually runs with TF_ROCM_FUSION_ENABLE and shows nice performance boost.

**Comment by sunway513** [comment]
Hi @TheCBaH , I can reproduce your observations. Please note TF_ROCM_FUSION_ENABLE is disabled by default, users should only enable it when it's known to work. We will refine the implementation of fusion on ROCm path to make it a more robust mechanism. 
Adding @deven-amd here to comment if the failure for the resnet/cifar10_main.py sample is expected with Fusion enabled.

**Comment by deven-amd** [comment]
Hi @TheCBaH ,

Can you try out the docker container : `devenamd/tensorflow:rocm2.4-tf1.13-python3-issue-456-fix`

It is based on the `` image, and contains the fix for the issue.

The bug seems to be that the "reserved" output ports for one of the "BatchNormGrad" nodes have connections on them. This is not expected by the code that creates the fusion nodes, and hence the error.

The "fix" is to merely detect this while we are determining fusion eligibility for a node, and skip fusion for nodes that fit this scenario.

I will also be filiing a PR to push this change to the `develop-upstream` branch

**Comment by TheCBaH** [comment]
Hi @deven-amd,

> Can you try out the docker container : `devenamd/tensorflow:rocm2.4-tf1.13-python3-issue-456-fix`

Yes,  `devenamd/tensorflow:rocm2.4-tf1.13-python3-issue-456-fix`  fixes the issue, thank you or quick turnaround.

>The "fix" is to merely detect this while we are determining fusion eligibility for a node, and skip fusion for nodes that fit this scenario.

Even, if fusion skipped for the batch normalization nodes, TF_ROCM_FUSION_ENABLE provides some boost for the entire network:
* without TF_ROCM_FUSION_ENABLE
```
I0518 05:42:44.440500 140737354032896 basic_session_run_hooks.py:680] global_step/sec: 14.024
I0518 05:42:44.441334 140737354032896 basic_session_run_hooks.py:247] cross_entropy = 1.934088, learning_rate = 0.1, train_accuracy = 0.1953125 (7.130 sec)
I0518 05:42:44.441634 140737354032896 basic_session_run_hooks.py:247] loss = 2.4139876, step = 200 (7.130 sec)
I0518 05:42:51.596385 140737354032896 basic_session_run_hooks.py:680] global_step/sec: 13.9746
I0518 05:42:51.597278 140737354032896 basic_session_run_hooks.py:247] cross_entropy = 1.8011167, learning_rate = 0.1, train_accuracy = 0.23242188 (7.156 sec)
I0518 05:42:51.597656 140737354032896 basic_session_run_hooks.py:247] loss = 2.2656488, step = 300 (7.156 sec)
I0518 05:42:58.753165 140737354032896 basic_session_run_hooks.py:680] global_step/sec: 13.9728
I0518 05:42:58.754460 140737354032896 basic_session_run_hooks.py:247] cross_entropy = 1.6648593, learning_rate = 0.1, train_accuracy = 0.2625 (7.157 sec)
I0518 05:42:58.754843 140737354032896 basic_session_run_hooks.py:247] loss = 2.1150098, step = 400 (7.157 sec)
I0518 05:43:05.922024 140737354032896 basic_session_run_hooks.py:680] global_step/sec: 13.9492
I0518 05:43:05.923305 140737354032896 basic_session_run_hooks.py:247] cross_entropy = 1.5323229, learning_rate = 0.1, train_accuracy = 0.29296875 (7.169 sec)
I0518 05:43:05.923877 140737354032896 basic_session_run_hooks.py:247] loss = 1.9692287, step = 500 (7.169 sec)
```
* with TF_ROCM_FUSION_ENABLE
```
I0518 05:44:23.737204 140737354032896 basic_session_run_hooks.py:247] cross_entropy = 1.9406083, learning_rate = 0.1, train_accuracy = 0.1625 (7.068 sec)
I0518 05:44:23.737712 140737354032896 basic_session_run_hooks.py:247] loss = 2.4758043, step = 400 (7.068 sec)
I0518 05:44:30.815824 140737354032896 basic_session_run_hooks.py:680] global_step/sec: 14.1244
I0518 05:44:30.817555 140737354032896 basic_session_run_hooks.py:247] cross_entropy = 1.8614838, learning_rate = 0.1, train_accuracy = 0.17578125 (7.080 sec)
I0518 05:44:30.818078 140737354032896 basic_session_run_hooks.py:247] loss = 2.378029, step = 500 (7.080 sec)
I0518 05:44:37.909114 140737354032896 basic_session_run_hooks.py:680] global_step/sec: 14.0978
I0518 05:44:37.910563 140737354032896 basic_session_run_hooks.py:247] cross_entropy = 1.7252854, learning_rate = 0.1, train_accuracy = 0.19642857 (7.093 sec)
I0518 05:44:37.911145 140737354032896 basic_session_run_hooks.py:247] loss = 2.2251897, step = 600 (7.093 sec)
I0518 05:44:44.992385 140737354032896 basic_session_run_hooks.py:680] global_step/sec: 14.1178
I0518 05:44:44.993691 140737354032896 basic_session_run_hooks.py:247] cross_entropy = 1.61365, learning_rate = 0.1, train_accuracy = 0.21484375 (7.083 sec)
I0518 05:44:44.994167 140737354032896 basic_session_run_hooks.py:247] loss = 2.097617, step = 700 (7.083 sec)
I0518 05:44:52.096229 140737354032896 basic_session_run_hooks.py:680] global_step/sec: 14.0767
```
